# Covid Data Analysis of Italy based on first six months of data(Time series data)

## By Saptaswa Basu

### Reading a CSV file

In [1]:
from urllib.request import urlretrieve

In [2]:
italy_covid_dataset = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'
urlretrieve(italy_covid_dataset,'italy-covid-beginningDays.csv')

('italy-covid-beginningDays.csv', <http.client.HTTPMessage at 0x262018543a0>)

In [3]:
!pip install pandas --upgrade --quiet

In [1]:
 import pandas as pd

In [2]:
covid_df = pd.read_csv('italy-covid-beginningDays.csv')

In [6]:
type(covid_df)

pandas.core.frame.DataFrame

In [6]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [8]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [7]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

In [11]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [15]:
covid_df.shape

(248, 4)

## Retrieving data fom a data frame
Conceptually dataframe is basically a dictionary of lists { [ ],[ ],...}, where 'keys' are 'columns' and 'values' are lists/arrays containing data for those respective columns.

In [10]:
covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

In [11]:
covid_df.new_cases

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

In [25]:
covid_df.at[246,'new_cases']

975.0

In [15]:
covid_df.loc[245]

date          2020-09-01
new_cases          996.0
new_deaths           6.0
new_tests        54395.0
Name: 245, dtype: object

To find the first valid indexes that doesnot contain a "NaN" value:--

In [17]:
covid_df.new_tests.first_valid_index()

111

In [18]:
covid_df.loc[108:114]

date  new_cases  new_deaths  new_tests
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0

In [21]:
covid_df.sample(10) 

date  new_cases  new_deaths  new_tests
194  2020-07-12      188.0         7.0    23061.0
213  2020-07-31      382.0         3.0    29686.0
127  2020-05-06     1075.0       236.0    37771.0
94   2020-04-03     4668.0       760.0        NaN
226  2020-08-13      476.0        10.0    25629.0
40   2020-02-09        0.0         0.0        NaN
209  2020-07-27      254.0         5.0    19374.0
19   2020-01-19        0.0         0.0        NaN
157  2020-06-05      177.0        88.0    40470.0
115  2020-04-24     2646.0       464.0    95273.0

In [45]:
covid_copy=covid_df.copy()

## Analyzing Data from data frames

In [23]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

#### Total number of reported cases and deaths based on the data:-

In [6]:
total_cases=sum(covid_df.new_cases)
total_deaths=sum(covid_df.new_deaths)

In [7]:
total_cases

271515.0

In [8]:
total_deaths

35497.0

#### Overall death ratio:-

In [9]:
death_rate=(total_deaths/total_cases)*100
death_rate

13.073679170579894

#### Overall test conducted (Around 935310 no. of test were conducted before these data's being reported):-

In [10]:
initial_test=935310
total_tests=initial_test+covid_df.new_tests.sum()
total_tests

5214766.0

#### Average ratio of positive results as per tests been conducted:-

In [11]:
positive_rate=(total_cases/total_tests)
positive_rate

0.05206657403227681

In [32]:
print('{:.2f}%'.format(positive_rate*100))

5.21%


## Querying and sorting rows of data

To check days which have more than 1000 reported cases

In [12]:
High_cases=covid_df.new_cases>1000
High_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

In [13]:
High_new_cases=covid_df[High_cases]
High_new_cases

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

From 248 rows it has come down to 72 rows

### Alternative

In [14]:
High_cases_df=covid_df[covid_df.new_cases>1000]
High_cases_df

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

Now above, only some rows are visible for faster outputs but sometimes we need to see all the rows of the datas maybe 100 or more. For that we can use 'display' functon with the 'option context' method

In [15]:
from IPython.display import display
with pd.option_context('display.max_rows',100):
    display(covid_df[covid_df.new_cases>1000])

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
74   2020-03-14     2547.0       252.0        NaN
75   2020-03-15     3497.0       173.0        NaN
76   2020-03-16     2823.0       370.0        NaN
77   2020-03-17     4000.0       347.0        NaN
78   2020-03-18     3526.0       347.0        NaN
79   2020-03-19     4207.0       473.0        NaN
80   2020-03-20     5322.0       429.0        NaN
81   2020-03-21     5986.0       625.0        NaN
82   2020-03-22     6557.0       795.0        NaN
83   2020-03-23     5560.0       649.0        NaN
84   2020-03-24     4789.0       601.0        NaN
85   2020-03-25     5249.0       743.0        NaN
86   2020-03-26     5210.0       685.0        NaN
87   2020-03-27     6153.0       660.0        NaN
88   2020-03-28     5959.0       971.0        NaN
89   2020-03-29     5974.0       887.0        NaN
90   2020-03-30     5217.0       758.0        NaN
91   2020-03-31     4050.0       810.0        NaN
92   2020-04-01     4053.0       839.0        NaN
93   2020-04-02     4782.0       727.0        NaN
94   2020-04-03     4668.0       760.0        NaN
95   2020-04-04     4585.0       764.0        NaN
96   2020-04-05     4805.0       681.0        NaN
97   2020-04-06     4316.0       527.0        NaN
98   2020-04-07     3599.0       636.0        NaN
99   2020-04-08     3039.0       604.0        NaN
100  2020-04-09     3836.0       540.0        NaN
101  2020-04-10     4204.0       612.0        NaN
102  2020-04-11     3951.0       570.0        NaN
103  2020-04-12     4694.0       619.0        NaN
104  2020-04-13     4092.0       431.0        NaN
105  2020-04-14     3153.0       564.0        NaN
106  2020-04-15     2972.0       604.0        NaN
107  2020-04-16     2667.0       578.0        NaN
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
115  2020-04-24     2646.0       464.0    95273.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
119  2020-04-28     1739.0       333.0    37554.0
120  2020-04-29     2091.0       382.0    38589.0
121  2020-04-30     2086.0       323.0    41441.0
122  2020-05-01     1872.0       285.0    43732.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
126  2020-05-05     1221.0       195.0    32211.0
127  2020-05-06     1075.0       236.0    37771.0
128  2020-05-07     1444.0       369.0    13665.0
129  2020-05-08     1401.0       274.0    45428.0
130  2020-05-09     1327.0       243.0    36091.0
131  2020-05-10     1083.0       194.0    31384.0
134  2020-05-13     1402.0       172.0    37049.0
236  2020-08-23     1071.0         3.0    47463.0
237  2020-08-24     1209.0         7.0    33358.0
240  2020-08-27     1366.0        13.0    57640.0
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

In [33]:
positive_rate

0.05206657403227681

In [16]:
High_positive_rate = covid_df[(covid_df.new_cases/covid_df.new_tests) > positive_rate]
High_positive_rate

date  new_cases  new_deaths  new_tests
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
120  2020-04-29     2091.0       382.0    38589.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
128  2020-05-07     1444.0       369.0    13665.0

In [17]:
covid_df['positive_rate'] = covid_df.new_cases/covid_df.new_tests

In [18]:
covid_df

date  new_cases  new_deaths  new_tests  positive_rate
0    2019-12-31        0.0         0.0        NaN            NaN
1    2020-01-01        0.0         0.0        NaN            NaN
2    2020-01-02        0.0         0.0        NaN            NaN
3    2020-01-03        0.0         0.0        NaN            NaN
4    2020-01-04        0.0         0.0        NaN            NaN
..          ...        ...         ...        ...            ...
243  2020-08-30     1444.0         1.0    53541.0       0.026970
244  2020-08-31     1365.0         4.0    42583.0       0.032055
245  2020-09-01      996.0         6.0    54395.0       0.018311
246  2020-09-02      975.0         8.0        NaN            NaN
247  2020-09-03     1326.0         6.0        NaN            NaN

[248 rows x 5 columns]

In [23]:
covid_df.drop(columns=['positive_rate'],inplace=True)

Here "inplace=True" helps to drop any column from existing variable, so that we dont need to initialize another variable to store the updated dataset.

In [24]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

## Sorting rows using column values

To find the days with highest number of cases & deaths:--

In [26]:
covid_df.sort_values('new_cases', ascending=False).head(10)

date  new_cases  new_deaths  new_tests
82  2020-03-22     6557.0       795.0        NaN
87  2020-03-27     6153.0       660.0        NaN
81  2020-03-21     5986.0       625.0        NaN
89  2020-03-29     5974.0       887.0        NaN
88  2020-03-28     5959.0       971.0        NaN
83  2020-03-23     5560.0       649.0        NaN
80  2020-03-20     5322.0       429.0        NaN
85  2020-03-25     5249.0       743.0        NaN
90  2020-03-30     5217.0       758.0        NaN
86  2020-03-26     5210.0       685.0        NaN

In [28]:
covid_df.sort_values('new_deaths', ascending=False).head(10)

date  new_cases  new_deaths  new_tests
88  2020-03-28     5959.0       971.0        NaN
89  2020-03-29     5974.0       887.0        NaN
92  2020-04-01     4053.0       839.0        NaN
91  2020-03-31     4050.0       810.0        NaN
82  2020-03-22     6557.0       795.0        NaN
95  2020-04-04     4585.0       764.0        NaN
94  2020-04-03     4668.0       760.0        NaN
90  2020-03-30     5217.0       758.0        NaN
85  2020-03-25     5249.0       743.0        NaN
93  2020-04-02     4782.0       727.0        NaN

From the peak days of cases and deaths, we can come to a conclusion that the peak of deaths started about a week after the peak in the daily new cases.

In [32]:
covid_df.sort_values('new_cases').head(10)

date  new_cases  new_deaths  new_tests
172  2020-06-20     -148.0        47.0    29875.0
0    2019-12-31        0.0         0.0        NaN
29   2020-01-29        0.0         0.0        NaN
30   2020-01-30        0.0         0.0        NaN
32   2020-02-01        0.0         0.0        NaN
33   2020-02-02        0.0         0.0        NaN
34   2020-02-03        0.0         0.0        NaN
36   2020-02-05        0.0         0.0        NaN
37   2020-02-06        0.0         0.0        NaN
38   2020-02-07        0.0         0.0        NaN

As we can see here at 172th row in the 'new_cases' gave a negetive data which might be a data entry error , now to resolve that the most useful approach we can follow is to replace the value with the avg of the value around it.    

In [33]:
covid_df.loc[168:175]

date  new_cases  new_deaths  new_tests
168  2020-06-16      301.0        26.0    27762.0
169  2020-06-17      210.0        34.0    33957.0
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20     -148.0        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

In [37]:
covid_df.at[172,'new_cases'] = (covid_df.at[171,'new_cases'] + covid_df.at[173,'new_cases'])/2

In [38]:
covid_df.loc[168:175]

date  new_cases  new_deaths  new_tests
168  2020-06-16      301.0        26.0    27762.0
169  2020-06-17      210.0        34.0    33957.0
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20      297.5        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

In [39]:
covid_df.sort_values('new_cases').head(10)

date  new_cases  new_deaths  new_tests
0   2019-12-31        0.0         0.0        NaN
28  2020-01-28        0.0         0.0        NaN
29  2020-01-29        0.0         0.0        NaN
30  2020-01-30        0.0         0.0        NaN
32  2020-02-01        0.0         0.0        NaN
33  2020-02-02        0.0         0.0        NaN
35  2020-02-04        0.0         0.0        NaN
36  2020-02-05        0.0         0.0        NaN
37  2020-02-06        0.0         0.0        NaN
38  2020-02-07        0.0         0.0        NaN

### Working with dates

As it is a time series dataset, we can break it down to specific months or specific weeks.

In [40]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

As we can see here the datatype of dates mentioned as 'object', so we need to convert the datatype into 'datetime' format, 
now to do that we used 'pd.to_datetime' method.

In [41]:
covid_df['date']=pd.to_datetime(covid_df.date)

In [42]:
covid_df.date

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

To get different parts of date into separate columns :---

In [47]:
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [48]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5
..         ...        ...         ...        ...   ...    ...  ...      ...
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3

[248 rows x 8 columns]

In [50]:
covid_df_may = covid_df[covid_df.month == 5]
covid_df_may

date  new_cases  new_deaths  new_tests  year  month  day  weekday
122 2020-05-01     1872.0       285.0    43732.0  2020      5    1        4
123 2020-05-02     1965.0       269.0    31231.0  2020      5    2        5
124 2020-05-03     1900.0       474.0    27047.0  2020      5    3        6
125 2020-05-04     1389.0       174.0    22999.0  2020      5    4        0
126 2020-05-05     1221.0       195.0    32211.0  2020      5    5        1
127 2020-05-06     1075.0       236.0    37771.0  2020      5    6        2
128 2020-05-07     1444.0       369.0    13665.0  2020      5    7        3
129 2020-05-08     1401.0       274.0    45428.0  2020      5    8        4
130 2020-05-09     1327.0       243.0    36091.0  2020      5    9        5
131 2020-05-10     1083.0       194.0    31384.0  2020      5   10        6
132 2020-05-11      802.0       165.0    25823.0  2020      5   11        0
133 2020-05-12      744.0       179.0    39620.0  2020      5   12        1
134 2020-05-13     1402.0       172.0    37049.0  2020      5   13        2
135 2020-05-14      888.0       195.0    41131.0  2020      5   14        3
136 2020-05-15      992.0       262.0    39027.0  2020      5   15        4
137 2020-05-16      789.0       242.0    40657.0  2020      5   16        5
138 2020-05-17      875.0       153.0    33505.0  2020      5   17        6
139 2020-05-18      675.0       145.0    26101.0  2020      5   18        0
140 2020-05-19      451.0        99.0    40226.0  2020      5   19        1
141 2020-05-20      813.0       162.0    38617.0  2020      5   20        2
142 2020-05-21      665.0       161.0    40644.0  2020      5   21        3
143 2020-05-22      642.0       156.0    42987.0  2020      5   22        4
144 2020-05-23      652.0       130.0    42579.0  2020      5   23        5
145 2020-05-24      669.0       119.0    34206.0  2020      5   24        6
146 2020-05-25      531.0        50.0    20676.0  2020      5   25        0
147 2020-05-26      300.0        92.0    33944.0  2020      5   26        1
148 2020-05-27      397.0        78.0    37299.0  2020      5   27        2
149 2020-05-28      584.0       117.0    39838.0  2020      5   28        3
150 2020-05-29      593.0        70.0    38233.0  2020      5   29        4
151 2020-05-30      516.0        87.0    36051.0  2020      5   30        5
152 2020-05-31      416.0       111.0    28948.0  2020      5   31        6